src/authn/Authenticator.ts

```diff
@@ -0,0 +1,88 @@
+import Libp2p from 'libp2p'
+import PeerId from 'peer-id'
+
+import { AuthRequest, AuthResponse } from './AuthRequests'
+import { AuthSender, ProductionAuthSender } from './AuthSender'
+import { PrivateKey } from '../crypto'
+
+const PROTO_AUTH = '/xmtplabs/xmtp-v1/clientauth/0.1.0-beta1'
+
+// AuthResult is the primary return type.
+export interface AuthResult {
+  isAuthenticated: boolean
+  errorString?: string
+}
+
+export type AuthOptions = {
+  // Specify a different sending mechanism for the authenticator to use
+  alternativeSender: AuthSender
+}
+
+/**
+  Authenticator securely provides a clients Identity to an XMTP node in order
+  to allow the sending of messages
+ */
+export default class Authenticator {
+  identityKey: PrivateKey
+  libp2p: Libp2p
+  walletAddress: string
+  private authState: Map<string, boolean> = new Map()
+  private sender: AuthSender
+
+  constructor(libp2p: Libp2p, identityKey: PrivateKey, sender: AuthSender) {
+    this.identityKey = identityKey
+    this.libp2p = libp2p
+    this.walletAddress = identityKey.publicKey.walletSignatureAddress()
+    this.sender = sender
+  }
+
+  static create(
+    libp2p: Libp2p,
+    identityKey: PrivateKey,
+    authOpts?: AuthOptions
+  ): Authenticator {
+    const sender = authOpts?.alternativeSender
```

Comment:
I was a little confused by this `alternativeSender` option at first too; why not just call it `sender`?